internal/controlplane/handlers_authz.go

```diff
@@ -632,23 +625,120 @@ func (s *Server) UpdateRole(ctx context.Context, req *minder.UpdateRoleRequest)
 	}
-	if flags.Bool(ctx, s.featureFlags, flags.IDPResolver) {
-		identity, err = s.idClient.Resolve(ctx, sub)
+	return nil, util.UserVisibleError(codes.InvalidArgument, "one of subject or email must be specified")
+}
+
+func (s *Server) updateInvite(
+	ctx context.Context,
+	targetProject uuid.UUID,
+	authzRole authz.Role,
+	email string,
+) (*minder.UpdateRoleResponse, error) {
+	var userInvite db.UserInvite
+	// Get the sponsor's user information (current user)
+	currentUser, err := s.store.GetUserBySubject(ctx, auth.GetUserSubjectFromContext(ctx))
+	if err != nil {
+		return nil, status.Errorf(codes.Internal, "failed to get user: %s", err)
+	}
+
+	// Get all invitations for this email and project
+	existingInvites, err := s.store.GetInvitationsByEmailAndProject(ctx, db.GetInvitationsByEmailAndProjectParams{
+		Email:   email,
+		Project: targetProject,
+	})
+	if err != nil {
+		return nil, status.Errorf(codes.Internal, "error getting invitations: %v", err)
+	}
+
+	// Exit early if there are no or multiple existing invitations for this email and project
+	if len(existingInvites) == 0 {
+		return nil, util.UserVisibleError(codes.NotFound, "no invitations found for this email and project")
+	} else if len(existingInvites) > 1 {
+		return nil, status.Errorf(codes.Internal, "multiple invitations found for this email and project")
+	}
+
+	// At this point, there should be exactly 1 invitation. We should either update its expiration or
+	// discard it and create a new one
+	if existingInvites[0].Role != authzRole.String() {
+		// If there's an existing invite with a different role, we should delete it and create a new one
+		// Delete the existing invitation
+		_, err = s.store.DeleteInvitation(ctx, existingInvites[0].Code)
 		if err != nil {
-			zerolog.Ctx(ctx).Error().Err(err).Msg("error resolving identity")
-			return nil, util.UserVisibleError(codes.NotFound, "could not find identity %q", sub)
+			return nil, status.Errorf(codes.Internal, "error deleting previous invitation: %v", err)
+		}
+		// Create a new invitation
+		userInvite, err = s.store.CreateInvitation(ctx, db.CreateInvitationParams{
+			Code:    invite.GenerateCode(),
+			Email:   email,
+			Role:    authzRole.String(),
+			Project: targetProject,
+			Sponsor: currentUser.ID,
+		})
+		if err != nil {
+			return nil, status.Errorf(codes.Internal, "error creating invitation: %v", err)
+		}
+	} else {
+		// If the role is the same, we should update the expiration
+		userInvite, err = s.store.UpdateInvitation(ctx, existingInvites[0].Code)
+		if err != nil {
+			return nil, status.Errorf(codes.Internal, "error updating invitation: %v", err)
 		}
 	}
+	// Resolve the project's display name
+	prj, err := s.store.GetProjectByID(ctx, userInvite.Project)
+	if err != nil {
+		return nil, status.Errorf(codes.Internal, "failed to get project: %s", err)
+	}
+	// Resolve the sponsor's identity and display name
+	identity, err := s.idClient.Resolve(ctx, currentUser.IdentitySubject)
```

Comment:
Should this be done earlier to avoid touching the DB if we can't resolve the identity (wrapping the db writes in a transaction would help, too)

Reply:
Addressed it in - 4eb2fe6674474acfb331ec1c1daeae83cad6d8a2